packages/fx-core/src/plugins/solution/fx-solution/arm.ts

```diff
@@ -0,0 +1,215 @@
+        if (result.value.Modules) {
+          for (const module of Object.entries(result.value.Modules)) {
+            const moduleFileName = module[0];
+            const moduleFileContent = module[1].Content;
+            moduleFiles.set(generateBicepModuleFilePath(moduleFileName), moduleFileContent);
+          }
+        }
+      } else {
+        const msg = format(getStrings().solution.ProvisionFailNotice, ctx.projectSettings?.appName);
+        ctx.logProvider?.error(msg);
+        return result;
+      }
+    }
+  }
+
+  // Write bicep content to project folder
+  if (bicepOrchestrationTemplate.needsGenerateTemplate()) {
+    // Output main.bicep file
+    const bicepOrchestrationFileContent =
+      bicepOrchestrationTemplate.renderOrchestrationFileContent();
+    const templateFolderPath = path.join(ctx.root, baseFolder, templateFolder);
+    await fs.ensureDir(templateFolderPath);
+    await fs.writeFile(
+      path.join(templateFolderPath, bicepOrchestrationFileName),
+      bicepOrchestrationFileContent
+    );
+
+    // Output bicep module files from each resource plugin
+    for (const module of moduleFiles) {
+      await fs.writeFile(path.join(templateFolderPath, module[0]), module[1]);
+    }
+
+    // Output parameter file
+    const parameterFileContent = bicepOrchestrationTemplate.getParameterFileContent();
+    const parameterFolderPath = path.join(ctx.root, baseFolder, parameterFolder);
+    await fs.ensureDir(parameterFolderPath);
+    await fs.writeFile(
+      path.join(parameterFolderPath, parameterTemplateFileName),
+      parameterFileContent
+    );
+  }
+
+  return ok(undefined); // Nothing to return when success
+}
+
+// Context used by handlebars to render the main.bicep file
+export class ArmTemplateRenderContext {
+  public Plugins: string[];
+  public PluginOutput: { [PluginName: string]: PluginOutputContext };
+
+  constructor(pluginNames: string[]) {
+    this.Plugins = pluginNames;
+    this.PluginOutput = {};
+  }
+
+  public addPluginOutput(pluginName: string, scaffoldResult: ScaffoldArmTemplateResult) {
+    const pluginOutputContext: PluginOutputContext = {
+      Modules: {},
+      Outputs: {},
+    };
+    const modules = scaffoldResult.Modules;
+    const outputs = scaffoldResult.Orchestration.ModuleTemplate?.Outputs;
+
+    if (modules) {
+      for (const module of Object.entries(modules)) {
+        const moduleFileName = module[0];
+        pluginOutputContext.Modules![moduleFileName] = {
+          Path: generateBicepModuleFilePath(moduleFileName),
+        };
+      }
+    }
+
+    if (outputs) {
+      for (const output of Object.entries(outputs)) {
+        const outputKey = output[0];
+        const outputValue = output[1];
+        pluginOutputContext.Outputs![outputKey] = outputValue;
+      }
+    }
+
+    this.PluginOutput[pluginName] = pluginOutputContext;
+  }
+}
+
+// Stores the bicep orchestration information for all resource plugins
+class BicepOrchestrationTemplate {
```

Comment:
This class has exactly the same name as the interface. Can we rename it to reduce confusion?

Reply:
thanks, renamed to BicepOrchestrationContent